Mathlib/Analysis/Complex/Hadamard.lean

```diff
@@ -0,0 +1,488 @@
+  exact HasBasis.inf_principal (HasBasis.comap _ (HasBasis.sup atBot_basis atTop_basis)) _
+
+lemma cocompact_strip.mem_sets {s : Set ℂ} : s ∈ cocompactStrip 0 1 ↔ ∃ T, ∀ (z : ℂ),
+    z.re ∈ Icc 0 (1 : ℝ) → T ≤ |z.im| → z ∈ s := by
+  rw [HasBasis.mem_iff cocompact_strip.basis]
+  constructor
+  · rintro ⟨⟨a,b⟩, h⟩
+    simp only [preimage_union, ge_iff_le, zero_le_one, not_true, gt_iff_lt, true_and] at h
+    use max (-a) b
+    intros z hset him
+    apply h
+    refine mem_inter ?_ hset
+    simp only [mem_preimage, mem_Iic, mem_Ici, mem_union]
+    simp only [abs_ofReal, max_le_iff] at him
+    cases' him with hima himb
+    by_cases 0 ≤ z.im
+    · right; rwa [← _root_.abs_of_nonneg h]
+    · left; rwa [← neg_le_neg_iff, ← abs_of_neg (not_le.mp h)]
+  · rintro ⟨T, h⟩
+    use (-T, T)
+    simp only [preimage_union, true_and]
+    rintro z ⟨him, hset⟩
+    apply h _ hset
+    rwa [le_abs, ← le_neg, or_comm, ← mem_Iic, ← mem_Ici,
+      ← mem_union, ← mem_preimage, preimage_union]
+
+lemma _root_.Real.tendsto_sq_cocompact_atTop : Tendsto (fun x ↦ x^2) (cocompact ℝ) atTop := by
+  convert_to Tendsto (fun x ↦ ‖x‖ * ‖x‖) (cocompact ℝ) atTop
+  · ext x; rw [pow_two]; exact (abs_mul_abs_self _).symm
+  · exact tendsto_norm_cocompact_atTop.atTop_mul_atTop tendsto_norm_cocompact_atTop
+
+lemma cocompact_strip.tendsto_sq_im_atTop : Tendsto (fun z : ℂ ↦ z.im^2)
+    (cocompactStrip 0 1) atTop := by
+  apply tendsto_inf_left _
+  change Tendsto ((fun x ↦ x^2) ∘ im) (comap im (cocompact ℝ)) atTop
+  rw [tendsto_comap'_iff]
+  · convert_to Tendsto (fun x ↦ ‖x‖ * ‖x‖) (cocompact ℝ) atTop
+    · ext x; simp only [Real.rpow_two, pow_two, Real.norm_eq_abs, abs_mul_abs_self]
+    · exact tendsto_norm_cocompact_atTop.atTop_mul_atTop tendsto_norm_cocompact_atTop
+  ·  simp only [univ_mem, range_im]
+
+lemma cocompact_strip.eventually_sq_re_le_one : ∀ᶠ (z : ℂ) in
+    (cocompactStrip 0 1), z.re^2 ≤ 1 := by
+  rw [cocompactStrip, comap_principal, eventually_inf_principal]
+  apply eventually_of_forall
+  rintro z hz
+  simp only [Real.rpow_two]
+  rw [sq_le_one_iff hz.1]
+  exact hz.2
+
+-- Smoothly decreasing function when the real part is bounded eventually ≤ 1
+lemma expterm_eventually_le_one (C : ℝ) (n : ℕ) (hn : 1 ≤ n) : ∀ᶠ (z : ℂ)
+    in (comap Complex.im (cocompact ℝ) ⊓ (comap Complex.re (principal  (Icc 0 1)))),
+    C * abs (exp ((z^(2 : ℕ)-1) * (n : ℝ)⁻¹) ) ≤ 1 := by
+  apply eventually_le_of_tendsto_lt  (zero_lt_one' ℝ) _
+  simp only [abs_exp]
+  rw [ show (fun z : ℂ ↦ C * Real.exp (((z^(2 : ℕ)-1) * (n : ℝ)⁻¹).re))
+    = (fun z : ℂ ↦ C * Real.exp ((z.re^(2 : ℕ) - z.im^(2 : ℕ) - 1) * (n : ℝ)⁻¹)) by
+      ext1 z
+      nth_rewrite 1 [← re_add_im z]
+      simp only [mul_re, sub_re, ofReal_im, mul_zero, tsub_zero, ofReal_re]
+      rw [cpow_nat_cast, sq, mul_re]
+      simp only [re_add_im, one_re, Nat.cast_ofNat, Real.rpow_two, mul_eq_mul_left_iff,
+      Real.exp_eq_exp, mul_eq_mul_right_iff, sub_left_inj, inv_eq_zero, Nat.cast_eq_zero, sq] ]
```

Comment:
That `show` should really be a `have` just before the `rw` call.

Reply:
Done